rust/datafusion/src/optimizer/boolean_comparison.rs

```diff
@@ -0,0 +1,270 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Boolean comparision rule rewrites redudant comparison expression involing boolean literal into
+//! unary expression.
+
+use std::sync::Arc;
+
+use crate::error::Result;
+use crate::logical_plan::{Expr, LogicalPlan, Operator};
+use crate::optimizer::optimizer::OptimizerRule;
+use crate::optimizer::utils;
+use crate::scalar::ScalarValue;
+
+/// Optimizer that simplifies comparison expressions involving boolean literals.
+///
+/// Recursively go through all expressionss and simplify the following cases:
+/// * `expr = ture` to `expr`
+/// * `expr = false` to `!expr`
+pub struct BooleanComparison {}
+
+impl BooleanComparison {
```

Comment:
I think we could make this optimizer a bit more generic (not necessary in this PR) to split recursion / pattern match

This is the more "optimizer framework" I had in mind in the comment on the roadmap
@alamb @jorgecarleitao @vertexclique .

A common strategy (used by Spark) for rule / replacement based  to have a loop that just does something like this:

```rust
let changed = False;

while !changed {
    (logical_plan, changed)  = apply_optimizations(rules, logical_plan);
}
```

A rule could be something like (returning `Some` on replaced output) and doesn't need the boilerplate of recursion for every rule.

```rust
// Optimizer can work both on expr and logical plans, default returns `None`
impl OptimizerRule for BooleanComparison {
   fn optimize_expr(&mut self, plan: &Expr) -> Option<Expr> {
       match e {
             Expr::BinaryExpr { left, op, right } => {
            let left = optimize_expr(left);
            let right = optimize_expr(right);
            match op {
                Operator::Eq => match (&left, &right) {
                    (Expr::Literal(ScalarValue::Boolean(b)), _) => match b {
                        Some(true) => Some(right),
                        Some(false) | None => Some(Expr::Not(Box::new(right))),
                    },
                    (_, Expr::Literal(ScalarValue::Boolean(b))) => match b {
                        Some(true) => Some(left),
                        Some(false) | None => Some(Expr::Not(Box::new(left))),
                    },
                    _ => None,
                },
                Operator::NotEq => match (&left, &right) {
                    (Expr::Literal(ScalarValue::Boolean(b)), _) => match b {
                        Some(false) | None => Some(right),
                        Some(true) => Some(Expr::Not(Box::new(right))),
                    },
                    (_, Expr::Literal(ScalarValue::Boolean(b))) => match b {
                        Some(false) | None => left,
                        Some(true) => Some(Expr::Not(Box::new(left))),
                    },
                    _ => None,
                },
                _ => None
            }
        }
       }
   }
}

```

Reply:
@Dandandan  -- I think this PR as written is quite efficient and doesn't need a convergence loop as you suggest (which I think ends up potentially being quite inefficient if many rewrites are required) -- it already does a depth first traversal of the tree, simplifying on the way up. 

I think convergence loops might be best used if we have several rules that can each potentially make changes that would unlock additional optimizations of the others

For example, if you had two different optimization functions like `optimization_A` and `optimization_B` but parts of `optimization_A` wouldn't be applied unless you ran `optimization_B`. In that case a loop like the following would let you take full advantage of that

```
while !changed {
  let exprA = optimization_A(expr);
  let exprB = optimization_B(exprA);
  changed = expr != exprB;
  expr = exprB
}
```

Reply:
As you can probably guess given PRs like https://github.com/apache/arrow/pull/9278 my preference to avoid repeating the structure walking logic is via a Visitor. Perhaps after this PR is merged, I can take a shot at rewriting it using a general `ExprRewriter` type pattern.

Reply:
:+1: Yes, agreed this ATM doesn't need a loop as there is no interaction yet with other rules. But I guess once you'll add them you will need it if you want to combine it with other rules. Using `expr != exprB` for `changed` might be a good way to start. The main thing I want to stress is that at some point we don't want the recursion itself in a rule, but in a more general "optimization framework".

The optimization loop itself can be written in such a way it does both top down and bottom up replacements, applying the same rule while the optimization "generates" a replacement for the node, so it doesn't need multiple traversals for cases like you mention. The concept is here in polars
https://github.com/ritchie46/polars/blob/master/polars/polars-lazy/src/logical_plan/optimizer/mod.rs#L69

And sounds like a perfect candidate to try for `ExprRewriter` :+1:

Reply:
yeah, totally, agree, i was really tempting to introduce an optimization framework while writing those tree traversal boilerplate code, but decided it's probably better be done as a separate refactoring so we can get a better feeling of how the new framework can affect all the existing rules. the way we are currently managing optimization rules is definitely too raw ;) 

I will take a look at the Expression Visitor pattern as well since that's something already exists in the code base.